src/python/pants/base/exception_sink.py

```diff
@@ -19,28 +24,181 @@
+
+  # These persistent open file descriptors are kept so faulthandler can figure out signal safety.
+  _pid_specific_error_fileobj = None
+  _shared_error_fileobj = None
+
+  # Integer code to exit with on an unhandled exception.
+  UNHANDLED_EXCEPTION_EXIT_CODE = 1
 
   def __new__(cls, *args, **kwargs):
     raise TypeError('Instances of {} are not allowed to be constructed!'
                     .format(cls.__name__))
 
+  class ExceptionSinkError(Exception): pass
+
+  # All reset_* methods are ~idempotent!
+  @classmethod
+  def reset_log_location(cls, new_log_location=None):
+    """
+    Class state:
+    - Overwrites `cls._log_dir`, `cls._pid_specific_error_fileobj`, and
+      `cls._shared_error_fileobj`.
+    OS state:
+    - May create a new directory.
+    - Overwrites signal handlers for many fatal signals.
+
+    :raises: :class:`ExceptionSink.ExceptionSinkError` if the directory is not writable.
+    """
+    # We could no-op here if the log locations are the same, but there's no reason not to have the
+    # additional safety of re-acquiring file descriptors each time.
+
+    if new_log_location:
+      # Create the directory if possible, or raise if not writable.
+      cls._check_or_create_new_destination(new_log_location)
+    else:
+      # Interpret a None argument to mean to just refresh the file handles for the current pid.
+      new_log_location = cls._log_dir
+
+    pid_specific_error_stream, shared_error_stream = cls._recapture_fatal_error_log_streams(
+      new_log_location)
+
+    # NB: mutate process-global state!
+    # Send a stacktrace to this file if interrupted by a fatal error.
+    faulthandler.enable(file=pid_specific_error_stream, all_threads=True)
+
+    # NB: mutate the class variables!
+    cls._log_dir = new_log_location
+    cls._pid_specific_error_fileobj = pid_specific_error_stream
+    cls._shared_error_fileobj = shared_error_stream
+
   @classmethod
-  def set_destination(cls, dir_path):
-    cls._destination = dir_path
+  def reset_exiter(cls, exiter):
+    """
+    Class state:
+    - Overwrites `cls._exiter`.
+    Python state:
+    - Overwrites sys.excepthook.
+    """
+    assert(isinstance(exiter, Exiter))
+    # NB: mutate the class variables! This is done before mutating the exception hook, because the
+    # uncaught exception handler uses cls._exiter to exit.
+    cls._exiter = exiter
+    # NB: mutate process-global state!
+    sys.excepthook = cls._log_unhandled_exception_and_exit
 
   @classmethod
-  def get_destination(cls):
-    return cls._destination
+  def reset_interactive_output_stream(cls, interactive_output_stream):
+    """
+    Class state:
+    - Overwrites `cls._interactive_output_stream`.
+    OS state:
+    - Overwrites the SIGUSR2 handler.
+
+    This is where the the error message on exit will be printed to as well.
+    """
+    # NB: mutate process-global state!
+    # TODO: chain=True will log tracebacks to previous values of the trace stream as well -- what
```

Comment:
The `faulthandler` docs indicate that you can call `unregister`or `disable` to remove previous copies of handlers. Is that worth doing here?

Reply:
At first I had it doing that, but ended up just copying the current code (hence the TODO, because writing to maybe-closed maybe-unrelated streams after explicitly changing it sounds like the wrong thing to do). I think it would make more sense to unregister and disable here.